src/charm.py

```diff
@@ -100,23 +110,86 @@ def _on_config_changed(self, event):
 
     def publish_config(self):
         """Generate config file and publish to all workers."""
-        mimir_config = self.coordinator.build_config(dict(self.config))
+        mimir_config = self.coordinator.build_config(self._s3_storage)
         self.cluster_provider.publish_configs(mimir_config)
 
+    def create_minio_buckets(self, conn: dict, bucket_names: list):
+        """Create Minio buckets."""
+        try:
+            client = Minio(
+                endpoint=f'{conn["service"]}.{conn["namespace"]}.svc.cluster.local:{conn["port"]}',
+                access_key=conn["access_key"],
+                secret_key=conn["secret_key"],
+                secure=conn["secure"],
+            )
+            for bucket in bucket_names:
+                found = client.bucket_exists(bucket)
+                if not found:
+                    client.make_bucket(bucket)
+                else:
+                    logger.info("Bucket already exists")
+        except S3Error:
+            logger.error("Error creating S3 buckets")
+
+    def _parse_s3_data(self, s3_data):
+        data_passed_dict = dict(item.split(": ") for item in s3_data.split("\n") if item)
+        data_passed_dict["secure"] = data_passed_dict.get("secure", "").lower() == "true"
+
+        return {
+            "access_key": data_passed_dict.get("access-key", ""),
+            "namespace": data_passed_dict.get("namespace", ""),
+            "port": data_passed_dict.get("port", ""),
+            "secret_key": data_passed_dict.get("secret-key", ""),
+            "secure": data_passed_dict.get("secure", ""),
+            "service": data_passed_dict.get("service", ""),
+        }
+
     def _on_mimir_cluster_changed(self, _):
-        if self.coordinator.is_coherent():
-            logger.info("mimir deployment coherent: publishing configs")
-            self.publish_config()
-        else:
-            logger.warning("this mimir deployment is incoherent")
+        if not self.coordinator.is_coherent():
+            logger.warning("Incoherent deployment: Some required Mimir roles are missing.")
+            return
+        self._process_s3_relation()
+        if not self._s3_storage_data and self.coordinator.is_scaled():
+            logger.warning("Filesystem storage cannot be used with replicated mimir workers")
+            return
+        self.publish_config()
+
+    def _process_s3_relation(self):
+        s3_relation = self.model.get_relation("s3")
+        if s3_relation and s3_relation.app:
+            data_passed = s3_relation.data.get(s3_relation.app, {}).get("data")
+            if data_passed:
```

Comment:
Could simplify with a walrus, 

```suggestion
            if data_passed := s3_relation.data.get(s3_relation.app, {}).get("data"):
```

but we should probably instantiate a pydantic model from the relation data, instead of fetching `.get("data")` ourselves here.